cpu/stm32/stmclk/stmclk_gx.c

```diff
@@ -191,73 +220,83 @@ void stmclk_init_sysclk(void)
-    /* if configured, we need to enable the HSE clock now */
-    RCC->CR |= RCC_CR_HSEON;
-    while (!(RCC->CR & RCC_CR_HSERDY)) {}
+        RCC->CFGR = (RCC_CFGR_SW_HSE | CLOCK_AHB_DIV | CLOCK_APB1_DIV | CLOCK_APB2_DIV);
 #endif
-    /* now we can safely configure and start the PLL */
-    RCC->PLLCFGR = (PLL_SRC | PLL_M | PLL_N | PLL_R | RCC_PLLCFGR_PLLREN);
-    RCC->CR |= RCC_CR_PLLON;
-    while (!(RCC->CR & RCC_CR_PLLRDY)) {}
-
-#if CLOCK_AHB > MHZ(80)
-    /* Divide HCLK by before enabling the PLL */
-    RCC->CFGR |= RCC_CFGR_HPRE_DIV2;
+        RCC->CFGR |= RCC_CFGR_SW_HSE;
+        while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_HSE) {}
+    }
+    else if (IS_ACTIVE(CONFIG_USE_CLOCK_PLL)) {
+#if defined(CPU_FAM_STM32G4)
+        if (CLOCK_AHB > MHZ(80)) {
+            /* Divide HCLK by 2 before enabling the PLL */
+            RCC->CFGR |= RCC_CFGR_HPRE_DIV2;
+        }
 #endif
 
-    /* now that the PLL is running, we use it as system clock */
-    RCC->CFGR |= RCC_CFGR_SW_PLL;
-    while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_PLL) {}
-
-#if CLOCK_AHB > MHZ(80)
-    /* Wait 1us before switching back to full speed */
-    /* Use volatile to prevent the compiler from optimizing the loop */
-    volatile uint8_t count = CLOCK_CORECLOCK / MHZ(1);
-    while (count--) {}
-    RCC->CFGR &= ~RCC_CFGR_HPRE_DIV2;
-#endif
+        /* now that the PLL is running, we use it as system clock */
+        RCC->CFGR |= RCC_CFGR_SW_PLL;
+        while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_PLL) {}
 
+#if defined(CPU_FAM_STM32G4)
+        if (CLOCK_AHB > MHZ(80)) {
+            /* Wait 1us before switching back to full speed */
+            /* Use volatile to prevent the compiler from optimizing the loop */
+            volatile uint8_t count = CLOCK_CORECLOCK / MHZ(1);
+            while (count--) {}
+            RCC->CFGR &= ~RCC_CFGR_HPRE_DIV2;
+        }
 #endif
+    }
 
-    stmclk_disable_hsi();
-    irq_restore(is);
+    if (!IS_ACTIVE(CLOCK_ENABLE_HSI)) {
+        /* Disable HSI only if not used */
+        stmclk_disable_hsi();
+    }
 
-#ifdef MODULE_PERIPH_HWRNG
-    /* HWRNG is clocked by HSI48 so enable this clock when the peripheral is used */
-    RCC->CRRCR |= RCC_CRRCR_HSI48ON;
-    while (!(RCC->CRRCR & RCC_CRRCR_HSI48RDY)) {}
+#if defined(CPU_FAM_STM32G4)
+    if (IS_USED(MODULE_PERIPH_HWRNG)) {
+        /* HWRNG is clocked by HSI48 so enable this clock when the peripheral is used */
+        RCC->CRRCR |= RCC_CRRCR_HSI48ON;
+        while (!(RCC->CRRCR & RCC_CRRCR_HSI48RDY)) {}
+    }
+
+    if (IS_USED(MODULE_PERIPH_RTT)) {
+        /* Ensure LPTIM1 clock source (LSI or LSE) is correctly reset when initializing
+           the clock, this is particularly useful after waking up from deep sleep */
+        if (IS_ACTIVE(CONFIG_BOARD_HAS_LSE)) {
+            RCC->CCIPR |= RCC_CCIPR_LPTIM1SEL_0 | RCC_CCIPR_LPTIM1SEL_1;
+        }
+        else {
+            RCC->CCIPR |= RCC_CCIPR_LPTIM1SEL_0;
+        }
+    }
```

Comment:
No RTT for G0?

Reply:
Interestingly, I found out that this piece of code is not needed: the RTT feature is still working, even after waking up from stop mode:

<details><summary>nucleo-g431rb</summary>

```
$ CFLAGS=-DPM_BLOCKER_INITIAL=0x01 make BOARD=nucleo-g431rb -C tests/periph_rtt flash test --no-print-directory
Building application "tests_periph_rtt" for "nucleo-g431rb" with MCU "stm32".

"make" -C /work/riot/RIOT/boards/nucleo-g431rb
"make" -C /work/riot/RIOT/boards/common/nucleo
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32/periph
"make" -C /work/riot/RIOT/cpu/stm32/stmclk
"make" -C /work/riot/RIOT/cpu/stm32/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  11564	    132	   2396	  14092	   370c	/work/riot/RIOT/tests/periph_rtt/bin/nucleo-g431rb/tests_periph_rtt.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/periph_rtt/bin/nucleo-g431rb/tests_periph_rtt.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01406-gca211373d-dirty (2020-09-29-10:54)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
srst_only separate srst_nogate srst_open_drain connect_assert_srst

Info : clock speed 2000 kHz
Info : STLINK V3J2M1 (API v3) VID:PID 0483:374E
Info : Target voltage: 3.323695
Info : stm32g4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32g4x.cpu on 0
Info : Listening on port 40775 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32g4x.cpu       hla_target little stm32g4x.cpu       reset

Info : Unable to match requested speed 2000 kHz, using 1000 kHz
Info : Unable to match requested speed 2000 kHz, using 1000 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x0800099c msp: 0x20000200
Info : device idcode = 0x20016468 (STM32G43/G44xx - Rev: Z)
Info : flash size = 128kbytes
Info : flash mode : single-bank
Warn : Adding extra erase range, 0x08002db0 .. 0x08002fff
auto erase enabled
wrote 11696 bytes from file /work/riot/RIOT/tests/periph_rtt/bin/nucleo-g431rb/tests_periph_rtt.elf in 0.460541s (24.801 KiB/s)

verified 11696 bytes in 0.160939s (70.970 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1000 kHz
Info : Unable to match requested speed 2000 kHz, using 1000 kHz
shutdown command invoked
Done flashing
r
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2021.01-devel-240-g070ea-pr/cpu/stm32gx_clock_cleanup)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
RTT now: 0
Setting initial alarm to now + 5 s (5120)
Done setting up the RTT, wait for many Hellos
Hello
Hello
Hello
Hello
Hello

```

</details>

<details><summary>nucleo-g071rb</summary>

```
$ CFLAGS=-DPM_BLOCKER_INITIAL=0x01 make BOARD=nucleo-g071rb -C tests/periph_rtt flash test --no-print-directory
Building application "tests_periph_rtt" for "nucleo-g071rb" with MCU "stm32".

"make" -C /work/riot/RIOT/boards/nucleo-g071rb
"make" -C /work/riot/RIOT/boards/common/nucleo
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32/periph
"make" -C /work/riot/RIOT/cpu/stm32/stmclk
"make" -C /work/riot/RIOT/cpu/stm32/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  10364	    132	   2396	  12892	   325c	/work/riot/RIOT/tests/periph_rtt/bin/nucleo-g071rb/tests_periph_rtt.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/periph_rtt/bin/nucleo-g071rb/tests_periph_rtt.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01406-gca211373d-dirty (2020-09-29-10:54)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
srst_only separate srst_nogate srst_open_drain connect_assert_srst

Info : clock speed 2000 kHz
Info : STLINK V2J31M21 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.217685
Info : stm32g0x.cpu: hardware has 4 breakpoints, 2 watchpoints
Info : starting gdb server for stm32g0x.cpu on 0
Info : Listening on port 36175 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32g0x.cpu       hla_target little stm32g0x.cpu       reset

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x080007a0 msp: 0x20000200
Info : device idcode = 0x20006460 (STM32G07/G08xx - Rev: B)
Info : flash size = 128kbytes
Info : flash mode : single-bank
Warn : Adding extra erase range, 0x08002900 .. 0x08002fff
auto erase enabled
wrote 10496 bytes from file /work/riot/RIOT/tests/periph_rtt/bin/nucleo-g071rb/tests_periph_rtt.elf in 0.319072s (32.124 KiB/s)

verified 10496 bytes in 0.080186s (127.828 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
shutdown command invoked
Done flashing
r
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
Help: Press s to start test, r to print it is ready
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2021.01-devel-240-g070ea-pr/cpu/stm32gx_clock_cleanup)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
RTT now: 0
Setting initial alarm to now + 5 s (5120)
Done setting up the RTT, wait for many Hellos
Hello
Hello
Hello
Hello
Hello

```

</details>

So I just dropped that part (for now). The CCIPR register management needs special care, especially when several bits must be set (for different peripherals). That could be done in follow-up PRs (there's the same issue on L4, I just noticed).